File: lib/cyanea/seq.ex
defmodule Cyanea.Seq do
  @moduledoc "Sequence validation, manipulation, pattern matching, and analysis."

  import Cyanea.NifHelper
  alias Cyanea.Native

  # ===========================================================================
  # Validation
  # ===========================================================================

  @type seq_type :: :dna | :rna | :protein

  @doc "Validate and normalize a sequence of the given type. Returns `{:ok, binary}` or `{:error, reason}`."
  @spec validate(binary(), seq_type()) :: {:ok, binary()} | {:error, term()}
  def validate(seq, :dna) when is_binary(seq), do: nif_call(fn -> Native.validate_dna(seq) end)
  def validate(seq, :rna) when is_binary(seq), do: nif_call(fn -> Native.validate_rna(seq) end)
  def validate(seq, :protein) when is_binary(seq), do: nif_call(fn -> Native.validate_protein(seq) end)

  @doc "Validate and normalize a sequence, raising on error."
  @spec validate!(binary(), seq_type()) :: binary()
  def validate!(seq, type) when is_binary(seq) do
    case validate(seq, type) do
      {:ok, result} -> result
      {:error, reason} -> raise ArgumentError, "validation failed: #{inspect(reason)}"
    end
  end

  # ===========================================================================
  # Operations
  # ===========================================================================

  @doc "Reverse complement a DNA sequence."
  @spec reverse_complement(binary()) :: {:ok, binary()} | {:error, term()}
  def reverse_complement(seq) when is_binary(seq),
    do: nif_call(fn -> Native.dna_reverse_complement(seq) end)

  @doc "Transcribe DNA to RNA (T -> U)."
  @spec transcribe(binary()) :: {:ok, binary()} | {:error, term()}
  def transcribe(seq) when is_binary(seq),
    do: nif_call(fn -> Native.dna_transcribe(seq) end)

  @doc "Translate RNA to protein (NCBI Table 1)."
  @spec translate(binary()) :: {:ok, binary()} | {:error, term()}
  def translate(seq) when is_binary(seq),
    do: nif_call(fn -> Native.rna_translate(seq) end)

  @doc "Calculate GC content of a DNA sequence (fraction 0.0-1.0)."
  @spec gc_content(binary()) :: {:ok, float()} | {:error, term()}
  def gc_content(seq) when is_binary(seq),
    do: nif_call(fn -> Native.dna_gc_content(seq) end)

  @doc "Extract k-mers from a DNA sequence."
  @spec kmers(binary(), pos_integer()) :: {:ok, [binary()]} | {:error, term()}
  def kmers(seq, k) when is_binary(seq) and is_integer(k),
    do: nif_call(fn -> Native.sequence_kmers(seq, k) end)

  @doc "Calculate molecular weight of a protein (Daltons)."
  @spec molecular_weight(binary()) :: {:ok, float()} | {:error, term()}
  def molecular_weight(seq) when is_binary(seq),
    do: nif_call(fn -> Native.protein_molecular_weight(seq) end)

  # ===========================================================================
  # Pattern matching
  # ===========================================================================

  @doc """
  Search for pattern matches in text.

  ## Options

    * `:algorithm` - `:horspool` (default, exact) or `:myers` (approximate)
    * `:max_distance` - maximum edit distance for Myers (default: 1)

  """
  @spec search(binary(), binary(), keyword()) :: {:ok, list()} | {:error, term()}
  def search(text, pattern, opts \\ [])

  def search(text, pattern, opts) when is_binary(text) and is_binary(pattern) do
    algorithm = Keyword.get(opts, :algorithm, :horspool)

    case algorithm do
      :horspool ->
        nif_call(fn -> Native.horspool_search(text, pattern) end)

      :myers ->
        max_dist = Keyword.get(opts, :max_distance, 1)
        nif_call(fn -> Native.myers_search(text, pattern, max_dist) end)
    end
  end

  # ===========================================================================
  # FM-Index
  # ===========================================================================

  @doc "Build an FM-index from text. Returns serialized index data."
  @spec build_index(binary()) :: {:ok, binary()} | {:error, term()}
  def build_index(text) when is_binary(text),
    do: nif_call(fn -> Native.fm_index_build(text) end)

  @doc "Count occurrences of pattern in FM-index."
  @spec count_occurrences(binary(), binary()) :: {:ok, non_neg_integer()} | {:error, term()}
  def count_occurrences(index_data, pattern) when is_binary(index_data) and is_binary(pattern),
    do: nif_call(fn -> Native.fm_index_count(index_data, pattern) end)

  # ===========================================================================
  # ORF finding
  # ===========================================================================

  @doc """
  Find open reading frames in both strands.

  ## Options

    * `:min_length` - minimum ORF length in nucleotides (default: 100)

  """
  @spec find_orfs(binary(), keyword()) :: {:ok, list()} | {:error, term()}
  def find_orfs(seq, opts \\ []) when is_binary(seq) do
    min_length = Keyword.get(opts, :min_length, 100)
    nif_call(fn -> Native.find_orfs(seq, min_length) end)
  end

  # ===========================================================================
  # MinHash
  # ===========================================================================

  @doc """
  Compute MinHash sketch of a sequence.

  ## Options

    * `:k` - k-mer size (default: 21)
    * `:sketch_size` - number of hash values (default: 1000)

  """
  @spec minhash(binary(), keyword()) :: {:ok, list()} | {:error, term()}
  def minhash(seq, opts \\ []) when is_binary(seq) do
    k = Keyword.get(opts, :k, 21)
    sketch_size = Keyword.get(opts, :sketch_size, 1000)
    nif_call(fn -> Native.minhash_sketch(seq, k, sketch_size) end)
  end

  @doc "Compute Jaccard similarity between two MinHash sketches."
  @spec minhash_jaccard(list(), list()) :: {:ok, float()} | {:error, term()}
  def minhash_jaccard(sketch_a, sketch_b) when is_list(sketch_a) and is_list(sketch_b),
    do: nif_call(fn -> Native.minhash_jaccard(sketch_a, sketch_b) end)

  # ===========================================================================
  # File I/O (sequence-specific)
  # ===========================================================================

  @doc "Get FASTA file statistics (sequence count, bases, GC content)."
  @spec fasta_stats(binary()) :: {:ok, struct()} | {:error, term()}
  def fasta_stats(path) when is_binary(path),
    do: nif_call(fn -> Native.fasta_stats(path) end)

  @doc "Get FASTQ file statistics (includes quality metrics)."
  @spec fastq_stats(binary()) :: {:ok, struct()} | {:error, term()}
  def fastq_stats(path) when is_binary(path),
    do: nif_call(fn -> Native.fastq_stats(path) end)

  @doc "Parse all records from a FASTQ file."
  @spec parse_fastq(binary()) :: {:ok, list()} | {:error, term()}
  def parse_fastq(path) when is_binary(path),
    do: nif_call(fn -> Native.parse_fastq(path) end)
end

File: native/cyanea_native/src/seq.rs
//! cyanea-seq NIFs — Sequence I/O, validation, operations, k-mers, pattern matching.

use crate::bridge::*;
use crate::to_nif_error;


// ===========================================================================
// Existing NIFs
// ===========================================================================

#[rustler::nif(schedule = "DirtyCpu")]
pub fn fasta_stats(path: String) -> Result<FastaStatsNif, String> {
    cyanea_seq::parse_fasta_stats(&path)
        .map(FastaStatsNif::from)
        .map_err(to_nif_error)
}

#[rustler::nif]
pub fn validate_dna(data: Vec<u8>) -> Result<Vec<u8>, String> {
    let seq = cyanea_seq::DnaSequence::new(&data).map_err(to_nif_error)?;
    Ok(seq.into_bytes())
}

#[rustler::nif]
pub fn validate_rna(data: Vec<u8>) -> Result<Vec<u8>, String> {
    let seq = cyanea_seq::RnaSequence::new(&data).map_err(to_nif_error)?;
    Ok(seq.into_bytes())
}

#[rustler::nif]
pub fn validate_protein(data: Vec<u8>) -> Result<Vec<u8>, String> {
    let seq = cyanea_seq::ProteinSequence::new(&data).map_err(to_nif_error)?;
    Ok(seq.into_bytes())
}

#[rustler::nif]
pub fn dna_reverse_complement(data: Vec<u8>) -> Result<Vec<u8>, String> {
    let seq = cyanea_seq::DnaSequence::new(&data).map_err(to_nif_error)?;
    Ok(seq.reverse_complement().into_bytes())
}

#[rustler::nif]
pub fn dna_transcribe(data: Vec<u8>) -> Result<Vec<u8>, String> {
    let seq = cyanea_seq::DnaSequence::new(&data).map_err(to_nif_error)?;
    Ok(seq.transcribe().into_bytes())
}

#[rustler::nif]
pub fn dna_gc_content(data: Vec<u8>) -> Result<f64, String> {
    let seq = cyanea_seq::DnaSequence::new(&data).map_err(to_nif_error)?;
    Ok(seq.gc_content())
}

#[rustler::nif]
pub fn rna_translate(data: Vec<u8>) -> Result<Vec<u8>, String> {
    let seq = cyanea_seq::RnaSequence::new(&data).map_err(to_nif_error)?;
    seq.translate()
        .map(|p| p.into_bytes())
        .map_err(to_nif_error)
}

#[rustler::nif]
pub fn sequence_kmers(data: Vec<u8>, k: usize) -> Result<Vec<Vec<u8>>, String> {
    let seq = cyanea_seq::DnaSequence::new(&data).map_err(to_nif_error)?;
    let kmers = seq
        .kmers(k)
        .map_err(to_nif_error)?
        .map(|kmer| kmer.to_vec())
        .collect();
    Ok(kmers)
}

#[rustler::nif(schedule = "DirtyCpu")]
pub fn parse_fastq(path: String) -> Result<Vec<FastqRecordNif>, String> {
    cyanea_seq::parse_fastq_file(&path)
        .map(|records| records.into_iter().map(FastqRecordNif::from).collect())
        .map_err(to_nif_error)
}

#[rustler::nif(schedule = "DirtyCpu")]
pub fn fastq_stats(path: String) -> Result<FastqStatsNif, String> {
    cyanea_seq::parse_fastq_stats(&path)
        .map(FastqStatsNif::from)
        .map_err(to_nif_error)
}

#[rustler::nif]
pub fn protein_molecular_weight(data: Vec<u8>) -> Result<f64, String> {
    let seq = cyanea_seq::ProteinSequence::new(&data).map_err(to_nif_error)?;
    Ok(seq.molecular_weight())
}

// ===========================================================================
// New NIFs
// ===========================================================================

#[rustler::nif]
pub fn horspool_search(text: Vec<u8>, pattern: Vec<u8>) -> Vec<usize> {
    cyanea_seq::horspool(&text, &pattern)
}

#[rustler::nif]
pub fn myers_search(text: Vec<u8>, pattern: Vec<u8>, max_dist: usize) -> Vec<(usize, usize)> {
    cyanea_seq::myers_bitparallel(&text, &pattern, max_dist)
}

#[rustler::nif(schedule = "DirtyCpu")]
pub fn fm_index_build(text: Vec<u8>) -> Vec<u8> {
    // Build to validate input, then return the raw text.
    // FmIndex doesn't implement Serialize, so we store the raw text and
    // rebuild the index lazily on each query.
    let _index = cyanea_seq::FmIndex::build(&text);
    text
}

#[rustler::nif]
pub fn fm_index_count(index_data: Vec<u8>, pattern: Vec<u8>) -> usize {
    let index = cyanea_seq::FmIndex::build(&index_data);
    index.count(&pattern)
}

#[rustler::nif]
pub fn find_orfs(seq: Vec<u8>, min_length: usize) -> Vec<OrfResultNif> {
    let results = cyanea_seq::find_orfs_both_strands(&seq, min_length);
    results
        .into_iter()
        .map(|orf| OrfResultNif {
            start: orf.start,
            end: orf.end,
            frame: orf.frame,
            strand: match orf.strand {
                cyanea_seq::Strand::Reverse => "-".into(),
                _ => "+".into(),
            },
            sequence: orf.sequence.clone(),
        })
        .collect()
}

#[rustler::nif]
pub fn minhash_sketch(seq: Vec<u8>, k: usize, sketch_size: usize) -> Result<Vec<u64>, String> {
    let mut mh = cyanea_seq::MinHash::new(k, sketch_size).map_err(to_nif_error)?;
    mh.add_sequence(&seq);
    Ok(mh.hashes().to_vec())
}
